wmget/wmget.1
.\"Generated by db2man.xsl. Don't modify this, modify the source.
.de Sh \" Subsection
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.TH "WMGET" 1 "" "" ""
.SH NAME
wmget \- Background download manager in a dockapp
.SH "SYNOPSIS"
.ad l
.hy 0
.HP 6
\fBwmget\fR dock [\fIoptions\fR]
.ad
.hy
.ad l
.hy 0
.HP 6
\fBwmget\fR [\fIoptions\fR] {\fIURL\fR}
.ad
.hy
.ad l
.hy 0
.HP 6
\fBwmget\fR cancel {\fIjob\-id\fR}
.ad
.hy
.ad l
.hy 0
.HP 6
\fBwmget\fR list
.ad
.hy

.SH "DESCRIPTION"

.PP
wmget is a ``dockapp'' which makes it more convenient to retrieve files in the background\&. Dockapps are applications which run in small windows intended to be ``docked'' in window manager\-provided locations\&. wmget was developed primarily under GNU Window Maker, the author's preferred WM, but is known to work under AfterStep as well, and should work with other dockapp\-aware window managers and docks\&.

.PP
It uses the excellent libcurl library, part of the Curl automated\-download program, to perform file retrieval from Web servers, FTP servers, and other sources\&.

.PP
wmget allows you to perform multiple downloads without keeping a terminal open (for FTP or curl or something) or another window on your desktop (e\&.g\&. for Mozilla download progress); download progress is visible any time the Dock is visible\&.

.PP
You start downloads either by ``pasting'' URLs from Web browsers or other applications, or by invoking wmget from the command line (or another script or program) with a source URL\&. The dockapp has a handful of configurable download options, such as target directory, HTTP proxy server, etc\&.

.SH "STARTING UP"

.PP
To start the dockapp, just run \fBwmget dock &\fR\&. If you are running Window Maker, you can then just drag the new appicon onto your Dock, right\-click on an area outside the four progress bars, select Settings, and select Start when Window Maker is started\&.

.PP
If you are running AfterStep, you can add it to your Wharf by adding the following line to your \fI~/GNUstep/Library/AfterStep/wharf\fR file:

.nf

        *Wharf wmget \- Swallow "wmget" wmget dock &
      
.fi

.PP
Other window managers support dockapps in different ways\&. Even in window managers without any special dockapp support, you can run wmget as noted above; it will simply show up as a small window or "icon"\&.

.SH "USING WMGET"

.PP
wmget's user interface is simple: four stacked progress bars, initially empty, representing four possible simultaneous downloads\&. The top bar will say ``wmget'' when there isn't a download running there, but any download will cover that up\&.

.PP
Each running download normally shows up to nine characters of its filename, overlaid with a progress bar\&. You can click on any progress bar to reveal a percentage display and a stop button; clicking on the percentage display switches back, while clicking on the stop button stops the download\&. There is currently no confirmation; it just stops\&.

.PP
You can ``request'' downloads at any time\&. If all four places show running downloads, additional requests will queue up, waiting for one to complete; wmget will never be downloading more than four files at a time\&.

.PP
By default, wmget figures out a reasonable filename for any requested downloads, writes them to your home directory, and won't overwrite an existing file by the same name\&. All of these, along with a few other options, are configurable\&. See below\&.

.SS "Requesting Downloads with the Mouse"

.PP
The easiest way to request a download is by copying and pasting a link\&. wmget lets you paste a URL by middle\-clicking anywhere on any of its status bars\&. Simply copy a link from some other source (for example, by right\-clicking on a link in Mozilla or Netscape and picking Copy Link Location), and middle\-click on one of the progress meter boxes in wmget\&.

.SS "Requesting Downloads from the Command Line"

.PP
The \fBwmget\fR command also lets you directly request downloads from the command line, or from within a script or another program\&. The syntax is \fBwmget \fIURL\fR\fR, plus any of the options documented below\&.

.PP
Once you run this command, you'll either get an error message or a ``job ID''\&. The job ID is only useful in conjunction with the \fBwmget cancel\fR command\&.

.SS "Download Failures"

.PP
Downloads can fail for a variety of reasons, from running out of disk space to modem hangups\&. Since wmget is designed not to interrupt your workflow or exceed its little square window, it responds to any download error by aborting the download and writing an error file to your download directory\&. This error file has the name \fIfile\&.ERROR\fR, where \fIfile\fR is the name of the actual download target\&. This error file is a plain text file containing information on what you were downloading and what went wrong\&.

.SS "Viewing and Canceling Downloads"

.PP
As noted above, you can see the currently\-running downloads in the four progress boxes on the dockapp\&. Clicking on a bar reveals a stop button, and clicking on that stop button cancels the download (but leaves the partially\-downloaded file on your computer)\&.

.PP
At any time, you can also run the \fBwmget list\fR command, which displays all the running downloads as well as any queued\-up requests\&. The listing contains entries like this:

.IP
.nf
Job 10 [linux\-2\&.6]: 1658544/33073407 RUNNING
ftp://ftp\&.kernel\&.org/pub/linux/kernel/v2\&.6/linux\-2\&.6\&.0\-test6\&.tar\&.bz2
=> /home/aaron/DOWNLOAD/linux\-2\&.6\&.0\-test6\&.tar\&.bz2

.fi

.PP
What you see in that (admittedly dense) listing are the job ID, the name of the download as displayed on the dockapp (surrounded in brackets), the progress in bytes, the total bytes to download, the current status, the source URL, and the target file on your computer\&. Whew\&.

.PP
You can cancel any requested or running download from the command line by specifying \fBwmget cancel \fIjob\-id\fR\fR\&.

.SH "COMMAND-LINE OPTIONS AND THE CONFIGURATION FILE"

.PP
wmget supports a handful of configuration options\&. You can specify defaults for all downloads by putting them in a configuration file or adding command\-line options to the \fBwmget dock\fR command at startup, or you can specify options for one specific download by adding options to the \fBwmget \fIURL\fR\fR command when you request them\&. There isn't any way to specify options on URLs you paste with the mouse\&. Dockapp command\-line arguments override config\-file settings, and per\-URL settings override dockapp settings\&.

.PP
The configuration file is an optional file named \fI\&.wmgetrc\fR in your home directory\&. If it's there, it's parsed by the dockapp at startup\&. The syntax is simple: one option per line, all options consisting of a name and possibly a value\&. Blank lines are okay, and lines starting with # are ignored (so you can disable options easily)\&. Option names are just the same as the command\-line option names given below, except you don't put the dashes (``\-\-'') and you can't use the one\-letter abbreviations\&.

.TP
\-\-version, \-v
Regardless of any other options, this prints out version and copyright information and exits\&.

.TP
\-\-help, \-h
Regardless of any other options, this prints out a help message and exits\&.

.TP
\-\-silent, \-s
Suppress any output text other than error messages\&.

.TP
\-\-verbose, \-V
Write extra debugging information; not very useful unless you're debugging or extending the software\&.

.TP
\-\-output \fIpathname\fR, \-o \fIpathname\fR
Specifies where to write downloaded files\&. In the config file or on the dockapp command line, this can only be used to specify your default download directory; it must be an existing directory, and if it's not absolute then it is assumed to be relative to your home directory\&. On a specific download request, this can provide an alternate save directory or even an alternate filename; in that case, a non\-absolute path is relative to the default download directory\&.

.TP
\-\-display \fIname\fR, \-d \fIname\fR
Display the first nine characters of \fIname\fR in the progress display for this file\&. (Only valid on specific download requests, not on the dockapp or in the config file\&.)

.TP
\-\-overwrite, \-O
Allow wmget to overwrite an existing file when downloading\&. Normally, it will refuse to do so\&.

.TP
\-\-continue, \-C
When fetching a file that already exists locally, assume the local copy was an aborted download and try to download just the remainder\&.

.TP
\-\-auth \fIusername\fR:\fIpassword\fR, \-a \fIusername\fR:\fIpassword\fR
Provides login information for the server from which you're downloading\&.

.TP
\-\-proxy \fIserver\fR:\fIport\fR, \-p \fIserver\fR:\fIport\fR, \-\-proxy_auth \fIuser\fR:\fIpassword\fR, \-P \fIuser\fR:\fIpassword\fR
Specifies a proxy server and optionally a proxy\-server username/password pair for getting past firewalls\&.

.TP
\-\-follow \fIN\fR, \-f \fIN\fR
Specifies how many HTTP redirects to follow when resolving a page; by default, wmget is configured to follow up to 5\&. Set this to 0 to disable redirection\&. (In any real\-world situation, if you're getting redirected more than 5 times, there's a problem\&.\&.\&.)

.TP
\-\-user\-agent \fIstring\fR, \-U \fIstring\fR
Specifies which User\-Agent string to provide to servers when performing HTTP downloads\&. The default User\-Agent names both the wmget and libcurl versions in use\&.

.TP
\-\-ascii, \-B
Force FTP downloads to use ASCII mode; normally, they use binary mode\&. If you're downloading text documents, ASCII mode will take care of any necessary conversions between the text formats of the server and your computer\&.

.TP
\-\-referer \fIstring\fR, \-e \fIstring\fR
Provides a ``referer'' string to the Web server\&.

.TP
\-\-interface \fIinterface\fR, \-\-n \fIinterface\fR
Names a specific network interface to use (e\&.g\&., eth0 for the first Ethernet interface on a Linux system)\&. Rarely needed\&.

.TP
\-\-headers, \-h
When performing an HTTP retrieval, include the HTTP message header in the saved file\&. This is only really useful for testing\&.

.SH "FILES"

.TP
\fI~/\&.wmgetrc\fR
The (optional) configuration file for the wmget dockapp\&. Settings in this file are used to specify defaults for the dockapp when it starts; see the section on configuration and command\-line options for more details\&.

.TP
\fI~/\&.wmget\&.iq\fR
A Unix\-domain socket created by the wmget dockapp to accept requests from wmget commands\&. Created at startup automatically\&.

.SH AUTHOR
Aaron Trickey.


wmget/wmget.c
/*
    wmget - A background download manager as a Window Maker dock app
    Copyright (c) 2001-2003 Aaron Trickey <aaron@amtrickey.net>

    Permission is hereby granted, free of charge, to any person
    obtaining a copy of this software and associated documentation files
    (the "Software"), to deal in the Software without restriction,
    including without limitation the rights to use, copy, modify, merge,
    publish, distribute, sublicense, and/or sell copies of the Software,
    and to permit persons to whom the Software is furnished to do so,
    subject to the following conditions:

    The above copyright notice and this permission notice shall be
    included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
    NONINFRINGEMENT.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

    ********************************************************************
    wmget.c - main() and a few common functions

    This file contains the main() for wmget; it simply checks for the
    presence of ``dock'' as the first argument and invokes server(),
    request(), cancel(), or list() as appropriate.  Also found here are
    a few global utilities.
*/

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <pwd.h>
#include <string.h>
#include <stdarg.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>

#include "wmget.h"

void debug_dump_job (Job *job)
{
    if (output_level () < OL_DEBUG)
        return;

    debug ("id = %lu\n"
           "status = %d\n"
           "progress=%d/%d\n"
           "stop_request=%d\n"
           "display=[%s]\n"
           "source_url=[%s]\n"
           "save_to=[%s]\n"
           "continue_from=%d\n\n",
           job->job_id,
           job->status, job->progress, job->stop_request,
           job->prog_max, job->options.display, job->source_url,
           job->options.save_to, job->options.continue_from);
}


const char *home_directory (void)
{
    static char *home_directory = 0;
    struct passwd *pwent;

    if (home_directory)
        return home_directory;

    home_directory = getenv ("HOME");

    if (!home_directory) {
        pwent = getpwuid (geteuid ());
        home_directory = strdup (pwent->pw_dir);
    }

    return home_directory;
}


int main (int argc, char **argv)
{
    char **arg;
    int (*function) (int, char **) = &request;

    if (argc < 2) {
        usage ();
        return 0;
    }

    /* Cheat, because we don't invoke getopt until we know what mode
     * we're in...
     */
    for (arg = argv; *arg; ++arg) {
        if (strcmp (*arg, "dock") == 0) {
            function = &server;

        } else if (strcmp (*arg, "cancel") == 0) {
            function = &cancel;

        } else if (strcmp (*arg, "list") == 0) {
            function = &list;

        } else if (strcmp (*arg, "-v") == 0
                || strcmp (*arg, "--version") == 0) {
            puts (WMGET_VERSION_BANNER);
            puts (WMGET_COPYRIGHT);
            return 0;

        } else if (strcmp (*arg, "-h") == 0
                || strcmp (*arg, "--help") == 0) {
            usage ();
            return 0;
        }
    }

    return function (argc, argv);
}






